js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataProvider.tsx

```diff
@@ -0,0 +1,292 @@
+import {
+  ASSETS_GRAPH_LIVE_QUERY,
+  ASSET_LATEST_INFO_FRAGMENT,
+  ASSET_NODE_LIVE_FRAGMENT,
+} from '../asset-graph/useLiveDataForAssetKeys';
+import {AssetKeyInput} from '../graphql/types';
+import {isDocumentVisible, useDocumentVisibility} from '../hooks/useDocumentVisibility';
+
+const _assetKeyListeners: Record<string, Array<DataForNodeListener>> = {};
+
+export function useAssetNodeLiveData(assetKeys: AssetKeyInput[]) {
+  const context = React.useContext(AssetLiveDataContext);
+  return context.useAssetNodeLiveData(assetKeys);
+}
+
+function _getAssetFromCache(client: ApolloClient<any>, uniqueId: string) {
+  const cachedAssetData = client.readFragment<AssetNodeLiveFragment>({
+    fragment: ASSET_NODE_LIVE_FRAGMENT,
+    fragmentName: 'AssetNodeLiveFragment',
+    id: `assetNodeLiveFragment-${uniqueId}`,
+  });
+  const cachedLatestInfo = client.readFragment<AssetLatestInfoFragment>({
+    fragment: ASSET_LATEST_INFO_FRAGMENT,
+    fragmentName: 'AssetLatestInfoFragment',
+    id: `assetLatestInfoFragment-${uniqueId}`,
+  });
+  if (cachedAssetData && cachedLatestInfo) {
+    return {cachedAssetData, cachedLatestInfo};
+  } else {
+    return null;
+  }
+}
+
+async function _queryAssetKeys(client: ApolloClient<any>, assetKeys: AssetKeyInput[]) {
+  const {data} = await client.query<AssetGraphLiveQuery, AssetGraphLiveQueryVariables>({
+    query: ASSETS_GRAPH_LIVE_QUERY,
+    fetchPolicy: 'network-only',
+    variables: {
+      assetKeys,
+    },
+  });
+  data.assetNodes.forEach((assetNode) => {
+    _assetKeyToUniqueId[JSON.stringify(assetNode.assetKey.path)] = assetNode.id;
+    client.writeFragment({
+      fragment: ASSET_NODE_LIVE_FRAGMENT,
+      fragmentName: 'AssetNodeLiveFragment',
+      id: `assetNodeLiveFragment-${assetNode.id}`,
+      data: assetNode,
+    });
+  });
+  data.assetsLatestInfo.forEach((assetLatestInfo) => {
+    _assetKeyToUniqueId[JSON.stringify(assetLatestInfo.assetKey.path)] = assetLatestInfo.id;
+    client.writeFragment({
+      fragment: ASSET_LATEST_INFO_FRAGMENT,
+      fragmentName: 'AssetLatestInfoFragment',
+      id: `assetLatestInfoFragment-${assetLatestInfo.id}`,
+      data: assetLatestInfo,
+    });
+  });
+}
+
+// How many assets to fetch at once
+const BATCH_SIZE = 50;
+
+// Milliseconds we wait until sending a batched query
+const BATCHING_INTERVAL = 100;
+
+export const SUBSCRIPTION_IDLE_POLL_RATE = 30 * 1000;
+// const SUBSCRIPTION_MAX_POLL_RATE = 2 * 1000;
+
+type DataForNodeListener = (stringKey: string, data: LiveDataForNode) => void;
+
+const AssetLiveDataContext = React.createContext<{
+  useAssetNodeLiveData: (keys: AssetKeyInput[]) => Record<string, LiveDataForNode | null>;
+}>({
+  useAssetNodeLiveData: () => ({}),
+});
+
+// Map of asset keys to their last fetched time and last requested time
+const lastFetchedOrRequested: Record<
+  string,
+  {fetched: number; requested?: undefined} | {requested: number; fetched?: undefined} | null
+> = {};
```

Comment:
It'd be ideal to have this state inside the context -- right now, the context is sort of a shim and the "real" context is here in global scope. I think it could make tests a bit tricky because you can't just remove the context and make a new fresh one, you also have to clear this hidden state to get a clean environment.

Reply:
I really wanted to avoid having this data in Context because I didn't want it to trigger any re-renders, I think it's fine for it to be global state. For testing we can always use `jest.resetModules()`